File: frontend/src/api/teachingSpace.ts
import api from "../utils/api";

export interface TeachingSpace {
  id: number;
  space_code: string | null;
  name: string;
  description: string | null;
  teacher_id: number;
  teacher_name: string | null;
  school_id: number | null;
  status: string;
  created_at: string;
  updated_at: string;
  major_count?: number;
  resource_count?: number;
  teacher_college_id?: number;
  teacher_college_name?: string;
  member_count?: number;
}

export interface SpaceMajor {
  id: number;
  space_id: number;
  major_id: number;
  major_name: string | null;
  major_code: string | null;
  enrollment_year: number | null;
  created_at: string;
}

export interface SpaceResources {
  vector_dbs: any[];
  bots: any[];
}

export const createTeachingSpace = async (name: string, description?: string): Promise<TeachingSpace> => {
  const formData = new FormData();
  formData.append('name', name);
  if (description) formData.append('description', description);
  const response = await api.post('/teaching-space', formData);
  return response.data.data;
};

export const listTeachingSpaces = async (): Promise<TeachingSpace[]> => {
  const response = await api.get('/teaching-space');
  return response.data.data;
};

export const getTeachingSpace = async (spaceId: number): Promise<TeachingSpace> => {
  const response = await api.get(`/teaching-space/${spaceId}`);
  return response.data.data;
};

export const updateTeachingSpace = async (spaceId: number, data: { name?: string; description?: string; space_status?: string }) => {
  const formData = new FormData();
  if (data.name) formData.append('name', data.name);
  if (data.description !== undefined) formData.append('description', data.description);
  if (data.space_status) formData.append('space_status', data.space_status);
  const response = await api.put(`/teaching-space/${spaceId}`, formData);
  return response.data.data;
};

export const deleteTeachingSpace = async (spaceId: number) => {
  const response = await api.delete(`/teaching-space/${spaceId}`);
  return response.data.data;
};

export const bindMajor = async (spaceId: number, majorId: number, enrollmentYear?: number | null): Promise<SpaceMajor> => {
  const formData = new FormData();
  formData.append('major_id', majorId.toString());
  if (enrollmentYear != null) formData.append('enrollment_year', enrollmentYear.toString());
  const response = await api.post(`/teaching-space/${spaceId}/bindMajor`, formData);
  return response.data.data;
};

export const unbindMajor = async (spaceId: number, majorId: number, enrollmentYear?: number | null) => {
  const params: Record<string, any> = {};
  if (enrollmentYear != null) params.enrollment_year = enrollmentYear;
  const response = await api.delete(`/teaching-space/${spaceId}/unbindMajor/${majorId}`, { params });
  return response.data.data;
};

export const getSpaceMajors = async (spaceId: number): Promise<SpaceMajor[]> => {
  const response = await api.get(`/teaching-space/${spaceId}/majors`);
  return response.data.data;
};

export const getSpaceStudents = async (spaceId: number): Promise<any[]> => {
  const response = await api.get(`/teaching-space/${spaceId}/students`);
  return response.data.data;
};

export const addSpaceStudent = async (spaceId: number, studentId: number) => {
  const formData = new FormData();
  formData.append('student_id', studentId.toString());
  const response = await api.post(`/teaching-space/${spaceId}/students`, formData);
  return response.data.data;
};

export const removeSpaceStudent = async (spaceId: number, studentId: number) => {
  const response = await api.delete(`/teaching-space/${spaceId}/students/${studentId}`);
  return response.data.data;
};

export const publishResource = async (spaceId: number, resourceType: string, resourceId: number) => {
  const formData = new FormData();
  formData.append('resource_type', resourceType);
  formData.append('resource_id', resourceId.toString());
  const response = await api.post(`/teaching-space/${spaceId}/publish`, formData);
  return response.data.data;
};

export const unpublishResource = async (spaceId: number, resourceType: string, resourceId: number) => {
  const formData = new FormData();
  formData.append('resource_type', resourceType);
  formData.append('resource_id', resourceId.toString());
  const response = await api.delete(`/teaching-space/${spaceId}/unpublish`, { data: formData });
  return response.data.data;
};

export const getSpaceResources = async (spaceId: number): Promise<SpaceResources> => {
  const response = await api.get(`/teaching-space/${spaceId}/resources`);
  return response.data.data;
};

export const listAllTeachingSpaces = async (organizationId?: number): Promise<TeachingSpace[]> => {
  const params: Record<string, any> = {};
  if (organizationId) params.organization_id = organizationId;
  const response = await api.get('/teaching-space/admin/all', { params });
  return response.data.data;
};

export const getSpacesByMajor = async (majorId: number): Promise<TeachingSpace[]> => {
  const response = await api.get(`/teaching-space/by-major/${majorId}`);
  return response.data.data;
};

File: backend/app/routers/teaching_space.py
"""
教学空间路由
"""
from fastapi import APIRouter, Depends, status, Form
from typing import Optional
from sqlalchemy.ext.asyncio import AsyncSession

from app.database_async import get_async_db
from app.utils.auth import get_current_user
from app.schemas.response import SuccessResponse
from app.models.user import User
from app.services.teaching_space_service import AsyncTeachingSpaceService
from app.utils.logger_config import get_logger
from app.utils.error_handler import ForbiddenError

logger = get_logger(__name__)
router = APIRouter()


@router.post("", status_code=status.HTTP_201_CREATED)
async def create_space(
    name: str = Form(...),
    description: Optional[str] = Form(None),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_async_db)
):
    """创建教学空间（仅老师）"""
    if current_user.role not in ("teacher", "admin"):
        raise ForbiddenError("仅教师可创建教学空间")
    result = await AsyncTeachingSpaceService.create_space(db, current_user, name, description)
    return SuccessResponse(message="创建成功", data=result)


@router.get("")
async def list_spaces(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_async_db)
):
    """获取教学空间列表（老师获取自己的，学生获取所在的）"""
    if current_user.role in ("teacher", "admin"):
        result = await AsyncTeachingSpaceService.list_spaces_for_teacher(db, current_user.id)
    else:
        result = await AsyncTeachingSpaceService.list_spaces_for_student(db, current_user.id)
    return SuccessResponse(message="查询成功", data=result)


@router.get("/admin/all")
async def list_all_spaces(
    organization_id: Optional[int] = None,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_async_db)
):
    """管理员获取所有教学空间（可按学院筛选）"""
    if current_user.role not in ("admin",):
        raise ForbiddenError("仅管理员可访问")
    result = await AsyncTeachingSpaceService.list_all_spaces(
        db, school_id=current_user.school_id, organization_id=organization_id
    )
    return SuccessResponse(message="查询成功", data=result)


@router.get("/by-major/{major_id}")
async def get_spaces_by_major(
    major_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_async_db)
):
    """获取某个专业绑定的所有教学空间"""
    result = await AsyncTeachingSpaceService.get_spaces_by_major(db, major_id)
    return SuccessResponse(message="查询成功", data=result)


@router.get("/{space_id}")
async def get_space(
    space_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_async_db)
):
    """获取教学空间详情"""
    result = await AsyncTeachingSpaceService.get_space(db, space_id)
    return SuccessResponse(message="查询成功", data=result)


@router.put("/{space_id}")
async def update_space(
    space_id: int,
    name: Optional[str] = Form(None),
    description: Optional[str] = Form(None),
    space_status: Optional[str] = Form(None),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_async_db)
):
    """更新教学空间"""
    result = await AsyncTeachingSpaceService.update_space(
        db, space_id, current_user.id, name, description, space_status, user_role=current_user.role
    )
    return SuccessResponse(message="更新成功", data=result)


@router.delete("/{space_id}", status_code=status.HTTP_200_OK)
async def delete_space(
    space_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_async_db)
):
    """删除教学空间"""
    await AsyncTeachingSpaceService.delete_space(db, space_id, current_user.id, user_role=current_user.role)
    return SuccessResponse(message="删除成功", data={"deleted": True})


@router.post("/{space_id}/bindMajor")
async def bind_major(
    space_id: int,
    major_id: int = Form(...),
    enrollment_year: Optional[int] = Form(None),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_async_db)
):
    """绑定专业到教学空间（可指定年级）"""
    result = await AsyncTeachingSpaceService.bind_major(
        db, space_id, major_id, current_user.id,
        user_role=current_user.role, enrollment_year=enrollment_year
    )
    return SuccessResponse(message="绑定成功", data=result)


@router.delete("/{space_id}/unbindMajor/{major_id}")
async def unbind_major(
    space_id: int,
    major_id: int,
    enrollment_year: Optional[int] = None,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_async_db)
):
    """解绑专业（可指定年级）"""
    await AsyncTeachingSpaceService.unbind_major(
        db, space_id, major_id, current_user.id,
        user_role=current_user.role, enrollment_year=enrollment_year
    )
    return SuccessResponse(message="解绑成功", data={"unbound": True})


@router.get("/{space_id}/majors")
async def get_majors(
    space_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_async_db)
):
    """获取空间绑定的专业列表"""
    result = await AsyncTeachingSpaceService.get_bound_majors(db, space_id)
    return SuccessResponse(message="查询成功", data=result)


@router.get("/{space_id}/students")
async def get_students(
    space_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_async_db)
):
    """获取空间内学生列表"""
    result = await AsyncTeachingSpaceService.get_space_students(db, space_id)
    return SuccessResponse(message="查询成功", data=result)


@router.post("/{space_id}/students")
async def add_student(
    space_id: int,
    student_id: int = Form(...),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_async_db)
):
    """添加学生到教学空间"""
    if current_user.role not in ("teacher", "admin"):
        raise ForbiddenError("仅教师或管理员可操作")
    result = await AsyncTeachingSpaceService.add_student(db, space_id, student_id, current_user.id, user_role=current_user.role)
    return SuccessResponse(message="添加成功", data=result)


@router.delete("/{space_id}/students/{student_id}")
async def remove_student(
    space_id: int,
    student_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_async_db)
):
    """从教学空间移除学生"""
    if current_user.role not in ("teacher", "admin"):
        raise ForbiddenError("仅教师或管理员可操作")
    await AsyncTeachingSpaceService.remove_student(db, space_id, student_id, current_user.id, user_role=current_user.role)
    return SuccessResponse(message="移除成功", data={"removed": True})


@router.post("/{space_id}/publish")
async def publish_resource(
    space_id: int,
    resource_type: str = Form(...),
    resource_id: int = Form(...),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_async_db)
):
    """分发资源到教学空间"""
    result = await AsyncTeachingSpaceService.publish_resource(
        db, space_id, current_user.id, resource_type, resource_id, user_role=current_user.role
    )
    return SuccessResponse(message="分发成功", data=result)


@router.delete("/{space_id}/unpublish")
async def unpublish_resource(
    space_id: int,
    resource_type: str = Form(...),
    resource_id: int = Form(...),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_async_db)
):
    """撤回资源分发"""
    await AsyncTeachingSpaceService.unpublish_resource(
        db, space_id, current_user.id, resource_type, resource_id, user_role=current_user.role
    )
    return SuccessResponse(message="撤回成功", data={"unpublished": True})


@router.get("/{space_id}/resources")
async def get_resources(
    space_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_async_db)
):
    """获取空间内的资源列表"""
    if current_user.role == "student":
        result = await AsyncTeachingSpaceService.get_student_space_resources(db, current_user.id, space_id)
    else:
        result = await AsyncTeachingSpaceService.get_space_resources(db, space_id)
    return SuccessResponse(message="查询成功", data=result)
